lib/note-list/decorators.js

```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+import replaceToArray from 'string-replace-to-array';
+
+import Checkbox from '../components/checkbox';
+import { taskPrefixRegex } from '../note-detail/toggle-task/constants';
+import randomString from '..//utils/crypto-random-string';
+
+const generateKey = () => randomString(6);
+
+export const decorateWith = (decorators, text) => {
+  let output = text;
+
+  decorators.forEach(({ filter, replacer }) => {
+    output = replaceToArray(output, filter, replacer);
+  });
+
+  return output;
+};
+
+export const checkboxDecorator = {
+  filter: taskPrefixRegex,
+  replacer: match => {
+    const isChecked = /x/i.test(match);
+    return <Checkbox checked={isChecked} key={generateKey()} />;
+  },
+};
+
+export const makeFilterDecorator = filter => ({
+  filter,
+  replacer: match => {
+    if (match.length) {
+      return (
+        <span className="search-match" key={generateKey()}>
```

Comment:
Good question! Both options are suboptimal 😬 I went this way because it makes the code a little more complex to keep track of the match indices here, and also because using array indices as component keys is generally bad practice.

But yeah, random keys aren’t great either — they’ll never cause weird reconciliation bugs, but will be less performant than an actual stable key.

I *think* we are safe to use indices here since we won't be doing any reordering, so let’s go that route 👍 https://github.com/Automattic/simplenote-electron/pull/1159/commits/5de6792bbf81f6c483043cd20135eb61df9be4c0